packages/hash/api/src/lib/schemas/jsonSchema.ts

```diff
@@ -41,57 +73,152 @@ export const generateSchema$id = (
   accountId: string,
   entityTypeId: string,
   relative: boolean = false,
-) => `${relative ? "" : FRONTEND_URL}/${accountId}/types/${entityTypeId}`;
+) => {
+  return `${relative ? "" : FRONTEND_URL}/${accountId}/types/${entityTypeId}`;
+};
 
 /**
  * Given a Schema$id, generate an appropriate $ref to put into allOf field on JSON schema.
  * This can be used to inherit from other schemas.
  * */
-export const schema$idRef = (schema$id: string) =>
-  JSON.stringify([{ $ref: schema$id }]);
+export const schema$idRef = (schema$id: string) => {
+  return JSON.stringify([{ $ref: schema$id }]);
+};
+
+type SchemaResolverFunction = (url: string) => Promise<AnySchemaObject>;
+
+const emptySchemaResolver = async (_url: string) => ({});
 
 /**
- * Create a JSON schema
- * @param title the name of the schema, e.g. Person
- * @param accountId the account it belongs to
- * @param entityTypeId the entityId of this entityType
- * @param maybeStringifiedSchema schema definition fields (in either a JSON string or JS object)
- *    (e.g. 'properties', 'definition', 'description')
- * @param description optional description for the type
- * @returns schema the complete JSON schema object
+ * Error that represents a type mismatch in the JSONSchema validation step.
+ * This error class exists for filtering purposes.
  */
-export const jsonSchema = async (
-  title: string,
-  accountId: string,
-  entityTypeId: string,
-  maybeStringifiedSchema: string | JSONObject = {},
-  description?: string,
-) => {
-  if (title[0] !== title[0].toUpperCase()) {
-    throw new Error(
-      `Schema title should be in PascalCase, you passed '${title}'`,
-    );
+export class TypeMismatch extends Error {
+  constructor(msg: string) {
+    super(msg);
+    Object.setPrototypeOf(this, TypeMismatch.prototype);
   }
+}
+
+/**
+ * Class that encapsulates JsonSchema validation.
+ */
+export class JsonSchemaCompiler extends DataSource {
+  private ajv!: Ajv2019;
+  private resolver!: SchemaResolverFunction;
```

Comment:
I don't think you need the `!` as you are setting these properties in the constructor.

Reply:
I think I went through TS docs too quick, as I thought this was used to define `lateinit` properties like in Kotlin. I see now that the types are not nullable in their signature without `!`. Thanks!